core/state_processor.go

```diff
@@ -159,7 +157,7 @@ func ApplyTransactionWithEVM(msg *Message, config *params.ChainConfig, gp *GasPo
 	// Update the state with pending changes.
 	var root []byte
 	if config.IsByzantium(blockNumber) {
-		tracingStateDB.Finalise(true)
+		evm.StateDB.Finalise(true)
```

Comment:
I think we should use tracingStateDB here. Finalise is hooked in some cases.

Reply:
The EVM is constructed with `tracingStateDB`.

But I agree it's a bit confusing here for have two stateDB instances, (a) the one in the EVM object (b) the supplied one for calling `IntermediateRoot` and making receipts.

The next step is to unify these two instances into a single one. The ideal status is we always use the evm.StateDB for interacting throughout the entire state transition.

Reply:
> The next step is to unify these two instances into a single one. The ideal status is we always use the evm.StateDB for interacting throughout the entire state transition.

Yeah. One way to do it would be, to instead of  `IntermediateState/Finalize`, have `func TxEnd(isEip158, isByzantium bool) common.Hash` and `BlockEnd(isEip158 bool) common.Hash`. And instead of 

```golang
			var root []byte
			if chainConfig.IsByzantium(vmContext.BlockNumber) {
				statedb.Finalise(true)
			} else {
				root = statedb.IntermediateRoot(chainConfig.IsEIP158(vmContext.BlockNumber)).Bytes()
			}
```
 we would do 
```
			isEip158 := chainConfig.IsEIP158(vmContext.BlockNumber)
			isByzantium := chainConfig.IsByzantium(vmContext.BlockNumber)
			root := statedb.TxEnd(isEip158, isByzantium)
```

--------------------------


As it is right _now_, we pass the statedb **struct** in order to 
1. Calculate root, via IntermediateRoot which is not exposed in `vm.StateDB` interface, 
2. To pass to receipt-building, which invokes the following methods on it

```
	if statedb.GetTrie().IsVerkle() {
		statedb.AccessEvents().Merge(evm.AccessEvents)
	}
```
and 
`statedb.GetLogs(tx.Hash(), blockNumber.Uint64(), blockHash)` and `statedb.TxIndex()`

So, in order to unify them into one, we need to do something clever about the receipt-building too. They way this PR works now is ok to me, as Gary pointed out -- the evm already uses the hookedstate, so `Finalise` does indeed operate on the hooked one.

Reply:
```
if statedb.GetTrie().IsVerkle() {
		statedb.AccessEvents().Merge(evm.AccessEvents)
	}
```

I would like to move the accessEvent into the EVM itself. So we can remove it from the statedb.